Core/src/org/sleuthkit/autopsy/geolocation/MapWaypoint.java

```diff
@@ -18,69 +18,229 @@
     }
-    
+
     /**
-     * Gets a list of jxmapviewer waypoints from the current case.
+     * Construct a Mapwaypoint from a GeoPosition
      * 
+     * @param position GeoPosition for the waypoint
+     */
+    private MapWaypoint(GeoPosition position) {
+        super(position.getLatitude(), position.getLongitude());
+        dataModelWaypoint = null;
+        this.position = position;
+    }
+
+    /**
+     * Gets a list of jxmapviewer waypoints from the current case.
+     *
      * @param skCase Current case
-     * 
+     *
      * @return List of jxmapviewer waypoints
-     * 
-     * @throws GeoLocationDataException 
+     *
+     * @throws GeoLocationDataException
      */
-    static List<org.jxmapviewer.viewer.Waypoint> getWaypoints(SleuthkitCase skCase) throws GeoLocationDataException{
+    static List<MapWaypoint> getWaypoints(SleuthkitCase skCase) throws GeoLocationDataException {
         List<Waypoint> points = Waypoint.getAllWaypoints(skCase);
-        
+
         List<Route> routes = Route.getRoutes(skCase);
-        for(Route route: routes) {
+        for (Route route : routes) {
             points.addAll(route.getRoute());
         }
-        
-        List<org.jxmapviewer.viewer.Waypoint> mapPoints = new ArrayList<>();
-        
-        for(Waypoint point: points) {
+
+        List<MapWaypoint> mapPoints = new ArrayList<>();
+
+        for (Waypoint point : points) {
             mapPoints.add(new MapWaypoint(point));
         }
-        
+
         return mapPoints;
     }
-    
+
+    /**
+     * Returns a MapWaypoint without a reference to the datamodel waypoint.
+     * 
+     * @param position Location for new waypoint
+     * 
+     * @return New MapWaypoint with dataModelWaypoint set to null
+     */
+    static MapWaypoint getDummyWaypoint(GeoPosition position) {
+        return new MapWaypoint(position);
+    }
+
     /**
      * {@inheritDoc}
      */
     @Override
     public GeoPosition getPosition() {
         return position;
     }
-    
+
+    /**
+     * Get the label for this waypoint
+     * 
+     * @return Waypoint label
+     */
     String getLabel() {
         return dataModelWaypoint.getLabel();
     }
+
+    /**
+     * Returns a list of JMenuItems for this Waypoint. When creating a menu the 
+     * list may contain nulls which should be removed or replaced with JSeparators
+     * 
+     * @return List of menu items
+     * 
+     * @throws TskCoreException 
+     */
+    JMenuItem[] getMenuItems() throws TskCoreException{
+        List<JMenuItem> menuItems = new ArrayList<>();
+        BlackboardArtifact artifact = dataModelWaypoint.getArtifact();
+        Content content = artifact.getSleuthkitCase().getContentById(artifact.getObjectID());
+
+        menuItems.addAll(getTimelineMenuItems(dataModelWaypoint.getArtifact()));
+        menuItems.addAll(getDataModelActionFactoryMenuItems(artifact, content));
```

Comment:
getTimelineMenuItems is getting called twice as it is also called in getDataModelActionFactoryMenuItems